src/box/lua/config/configdata.lua

```diff
@@ -246,6 +253,167 @@ local function effective_snapshot_dir(iconfig)
+        end
+    end
+    return nil
+end
+
+-- Validate UUIDs and names passed to config against the data,
+-- saved inside snapshot. Fail early and don't recover all
+-- data if mismatch is found. Returns a map of names, missing
+-- in snapshot in the following format: {<name> = <uuid>}
+local function validate_names(snap_names, config_names)
+    -- Snapshot always has replicaset uuid and at least one
+    -- peer in _cluster space.
+    assert(snap_names.replicaset_uuid)
+    assert(snap_names.instance_uuid)
+    assert(#snap_names.peers > 0)
+    -- Config always has names set.
+    assert(config_names.replicaset_name ~= nil)
+    assert(config_names.instance_name ~= nil)
+
+    if config_names.replicaset_uuid ~= nil and
+       config_names.replicaset_uuid ~= snap_names.replicaset_uuid then
+        error(string.format('Replicaset UUID mismatch. Snapshot: %s, ' ..
+                            'config: %s.', snap_names.replicaset_uuid,
+                            config_names.replicaset_uuid))
+    end
+
+    if snap_names.replicaset_name ~= nil and
+       snap_names.replicaset_name ~= config_names.replicaset_name then
+        error(string.format('Replicaset name mismatch. Snapshot: %s, ' ..
+                            'config: %s.', snap_names.replicaset_name,
+                            config_names.replicaset_name))
+    end
+
+    if config_names.instance_uuid ~= nil and
+       config_names.instance_uuid ~= snap_names.instance_uuid then
+        error(string.format('Instance UUID mismatch. Snapshot: %s, ' ..
+                            'config: %s.', snap_names.instance_uuid,
+                            config_names.instance_uuid))
+    end
+
+    if snap_names.instance_name ~= nil and
+       snap_names.instance_name ~= config_names.instance_name then
+        error(string.format('Instance name mismatch. Snapshot: %s, ' ..
+                            'config: %s.', snap_names.instance_name,
+                            config_names.instance_name))
+    end
+
+    -- Fail early, if current UUID is not set, but no name is found
+    -- inside the snapshot file.
+    --
+    -- TODO: it's NoOp for now, as instance_uuid is automatically
+    -- generated for vshard configuration. This error and error_uuids
+    -- can't happen and consequently not tested. We must get rid of
+    -- generating UUIDs and test it properly.
+    --
+    -- For now it's impossible to recover from snaps without manually
+    -- specifying UUIDs.
+    if snap_names.instance_name == nil and
+       config_names.instance_uuid == nil then
+        error(string.format('Instance name is not set in snapshot and UUID ' ..
+                            'is missing in the config. Found %s in snapshot.',
+                            snap_names.instance_uuid))
+    end
+
+    -- Check, that every peer, which doesn't have name in snap,
+    -- has UUID in config, so that we can validate, that the right
+    -- config is applied to the instance.
```

Comment:
I think that we shouldn't forbid to attach a non-anonymous replica manually without an instance in the cluster configuration. Moreover, I think that it should be possible to join such a replica with and without persisting its name.

I see several possible use cases.

1. A replicator/CDC tool that is written on Go or another language -- not a tarantool instance. It is possible that it doesn't support joining a replicaset with an instance name. However, it is possible that it joins with a name.
2. Low-level administration tasks. Say, one may need to attach a temporary instance with specific application specific compatibility logic during an upgrade.